### PR TITLE
Add UUID functions

### LINKED ISSUE #1013: Add UUID generation function and per-server UUIDs
We should be able to generate [version 4 UUIDs](https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_.28random.29) as strings fairly easily. Assigning each server a persistent UUID means per-server storage for E2 chips would be easy.

Originally mentioned in #1010.

### PR DESCRIPTION
**Pull request by AbigailBuccaneer**
Fixes #1013.

* `WireLib.GenerateUUID()` generates a new UUID
* `WireLib.GetServerUUID()` returns a persistent UUID associated with the server
* `e2function string serverUUID()` does the same as above

### CODE REVIEW

**Comment by oldmud0** [comment]
looks ok

**Comment by thegrb93** [comment]
Then the server deleted its cookies and isn't identified as the same server anymore.

**Comment by oldmud0** [comment]
what does that mean for E2s though?

**Comment by bigdogmat** [comment]
The server generates a new UUID and any E2 relying on the old one won't work anymore.

**Comment by Divran** [comment]
nothing we can do about that though

**Comment by bigdogmat** [comment]
@Divran Oh I know, I was just explaining for @oldmud0

**Comment by AbigailBuccaneer** [comment]
Does this look okay to everybody? @TomyLobo

File: lua/wire/wireshared.lua [hunk 1]
```diff
@@ -369,6 +369,45 @@ function WireLib.ErrorNoHalt(message)
...
+	-- MSVC's RAND_MAX = 0x7FFF, which means math.random(0, 0xFFFF) won't
+	-- return all possible values.
+	local bytes = {}
+	for i = 1, 16 do bytes[i] = math.random(0, 0xFF) end
+	bytes[7] = bit.bor(0x40, bit.band(bytes[7], 0x0F))
+	bytes[9] = bit.bor(0x80, bit.band(bytes[7], 0x3F))
+	return string.format("%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x", unpack(bytes))
+end
+
+local UUID_MESSAGE_NAME = "WireLib.GetServerUUID"
```
Comment (TomyLobo) [inline lua/wire/wireshared.lua]: Why is it called message?

File: lua/wire/wireshared.lua [hunk 2]
```diff
@@ -369,6 +369,45 @@ function WireLib.ErrorNoHalt(message)
...
+			uuid = WireLib.GenerateUUID()
+			cookie.Set(UUID_MESSAGE_NAME, uuid)
+		end
+		return uuid
+	end
+
+	util.AddNetworkString(UUID_MESSAGE_NAME)
+
+	hook.Add("PlayerInitialSpawn", UUID_MESSAGE_NAME, function(player)
+		net.Start(UUID_MESSAGE_NAME)
```
Comment (TomyLobo) [inline lua/wire/wireshared.lua]: Oh never mind the previous comment...
But should we really send stuff we don't need (yet)?
Also, I think @Divran told me umsg is better for tiny messages like this.

File: lua/wire/wireshared.lua [hunk 3]
```diff
@@ -588,7 +627,7 @@ if SERVER then
 	end
 
 	local function FlushQueue(lqueue, ply)
-		// Zero these two for the writemsg function
+		-- Zero these two for the writemsg function
```
Comment (TomyLobo) [inline lua/wire/wireshared.lua]: unrelated change is unrelated

File: lua/wire/wireshared.lua (hunk 1 above)
Comment (AbigailBuccaneer) [inline lua/wire/wireshared.lua]: Because it's the name used for the net message :wink:

File: lua/wire/wireshared.lua (hunk 2 above)
Comment (Divran) [inline lua/wire/wireshared.lua]: nah I think I only said that it doesn't matter when the message is tiny

Comment (TomyLobo) [inline lua/wire/wireshared.lua]: ah i see

File: lua/wire/wireshared.lua [hunk 4]
```diff
@@ -325,7 +325,7 @@ do
 		util.AddNetworkString("wire_addnotify")
 		function WireLib.AddNotify(ply, Message, Type, Duration, Sound)
 			if isstring(ply) then ply, Message, Type, Duration, Sound = nil, ply, Message, Type, Duration end
-			if ply && !ply:IsValid() then return end
+			if ply and not IsValid(ply) then return end
```
Comment (TomyLobo) [inline lua/wire/wireshared.lua]: Doesn't IsValid(ent) already check whether ent is nil?

File: lua/wire/wireshared.lua (hunk 1 above)
Comment (AbigailBuccaneer) [inline lua/wire/wireshared.lua]: Renamed to `PERSISTENT_UUID_KEY` which better reflects that it's used in multiple ways.

**Comment by AbigailBuccaneer** [comment]
@TomyLobo okay to merge?

File: lua/entities/gmod_wire_expression2/core/serverinfo.lua [hunk 5]
```diff
@@ -20,6 +20,10 @@ e2function string gamemode()
 	return gmod.GetGamemode().Name
 end
 
+e2function string serverUUID()
```
Comment (TomyLobo) [inline lua/entities/gmod_wire_expression2/core/serverinfo.lua]: This function is too long and needs to be refactored into smaller components

Comment (immibis) [inline lua/entities/gmod_wire_expression2/core/serverinfo.lua]: Is that sarcastic?

Comment (TomyLobo) [inline lua/entities/gmod_wire_expression2/core/serverinfo.lua]: yes
